mlir/lib/Analysis/Presburger/Printer.cpp

```diff
@@ -0,0 +1,169 @@
+  printVariableList(os, set.getNumDims(), set.getNumSyms());
+  os << " : ";
+  if (set.isMarkedEmpty()) {
+    os << "(1 = 0)";
+    return;
+  }
+  printConstraints(os, set.getFlatAffineConstraints());
+}
+
+void PresburgerPrinter::print(raw_ostream &os, const PresburgerExpr &expr) {
+  unsigned nDim = expr.getNumDims(), nSym = expr.getNumSyms();
+  printVariableList(os, nDim, nSym);
+
+  os << " -> ";
+
+  for (unsigned i = 0, e = expr.getExprs().size(); i < e; ++i) {
+    if (i != 0)
+      os << " ; ";
+
+    os << "(";
+    // TODO change this as soon as we have a Constraint class. Try to unify the
+    // handling of expression
+    auto eI = expr.getExprs()[i];
+    printExpr(os, eI.second, eI.first, nDim);
+    os << ")";
+    os << " : ";
+    printConstraints(os, expr.getDomains()[i].getFlatAffineConstraints());
+  }
+}
+
+/// Prints the '(d0, ..., dN)[s0, ... ,sM]' dimension and symbol list.
+///
+void PresburgerPrinter::printVariableList(raw_ostream &os, unsigned nDim,
+                                          unsigned nSym) {
+  os << "(";
+  for (unsigned i = 0; i < nDim; i++)
+    os << (i != 0 ? ", " : "") << 'd' << i;
+  os << ")";
+
+  if (nSym > 0) {
+    os << "[";
+    for (unsigned i = 0; i < nSym; i++)
+      os << (i != 0 ? ", " : "") << 's' << i;
+    os << "]";
+  }
+}
+
+/// Prints the constraints of each `FlatAffineConstraints`.
+///
+void PresburgerPrinter::printConstraints(
+    raw_ostream &os,
+    const SmallVectorImpl<FlatAffineConstraints> &flatAffineConstraints) {
+  os << "(";
+  bool fst = true;
+  for (auto &c : flatAffineConstraints) {
+    if (fst)
+      fst = false;
+    else
+      os << " or ";
+    printFlatAffineConstraints(os, c);
+  }
+  os << ")";
+}
+
+/// Prints the constraints of the `FlatAffineConstraints`. Each constraint is
+/// printed separately and the are conjuncted with 'and'.
+///
+void PresburgerPrinter::printFlatAffineConstraints(
+    raw_ostream &os, const FlatAffineConstraints &cs) {
+  unsigned numIds = cs.getNumIds();
+  for (unsigned i = 0, e = cs.getNumEqualities(); i < e; ++i) {
+    if (i != 0)
+      os << " and ";
+    ArrayRef<int64_t> eq = cs.getEquality(i);
+    printExpr(os, eq.take_front(numIds), eq[numIds], cs.getNumDimIds());
+    os << " = 0";
+  }
+
+  if (cs.getNumEqualities() > 0 && cs.getNumInequalities() > 0)
+    os << " and ";
+
+  for (unsigned i = 0, e = cs.getNumInequalities(); i < e; ++i) {
+    if (i != 0)
+      os << " and ";
+    ArrayRef<int64_t> ineq = cs.getInequality(i);
+    printExpr(os, ineq.take_front(numIds), ineq[numIds], cs.getNumDimIds());
+    os << " >= 0";
+  }
+}
+
+/// Prints the coefficient of the i'th variable with an additional '+' or '-' is
+/// first = false. First indicates if this is the first summand of an
+/// expression.
+///
+/// Returns failure if the coefficient value is 0 and therefore is not printed.
+///
+LogicalResult PresburgerPrinter::printCoef(raw_ostream &os, int64_t val,
```

Comment:
nit: I feel `coeff` is a more standard short form for "coefficient"

Reply:
I feel that not printing zero coefficients does not really indicate a _failure_ per se, hence using a bool may be more appropriate for the return type.

Reply:
I wasn't sure how they prefer it in LLVM/MLIR,i thought that it was somewhere stated that boolean returns can be confusing. But in this case I agree.